File: keystroke-data-collector/frontend/src/utils/keystrokeFeatures.js
/**
 * Keystroke Dynamics Feature Extraction
 * 
 * Comprehensive keystroke feature analysis including:
 * - Key hold time and flight time metrics
 * - Typing speed and rhythm indicators
 * - Pause duration and frequency
 * - Error and correction-related features
 * 
 * Features are normalized against participant-specific baselines
 */

/**
 * Extract keystroke metrics from raw keystroke events
 * @param {Array} keystrokesData - Array of keystroke events with timestamp
 * @param {string} typedText - The actual text typed by user
 * @param {string} promptText - The target text to type
 * @returns {Object} Comprehensive keystroke metrics
 */
export function extractKeystrokeFeatures(keystrokesData, typedText, promptText) {
  if (!keystrokesData || keystrokesData.length === 0) {
    return getEmptyFeatures();
  }

  // Core metrics
  const metrics = {
    // Key Hold Time Metrics
    keyHoldTimes: calculateKeyHoldTimes(keystrokesData),
    
    // Flight Time Metrics (inter-keystroke interval)
    flightTimes: calculateFlightTimes(keystrokesData),
    
    // Typing Speed Indicators
    typingSpeed: calculateTypingSpeed(keystrokesData, typedText),
    
    // Rhythm and Regularity
    rhythmMetrics: calculateRhythmMetrics(keystrokesData),
    
    // Pause Analysis
    pauses: analyzePauses(keystrokesData),
    
    // Error and Correction Features
    errorMetrics: calculateErrorMetrics(typedText, promptText, keystrokesData),
    
    // Keystroke Statistics
    keystrokeStats: calculateKeystrokeStats(keystrokesData, typedText),
  };

  return metrics;
}

/**
 * Calculate key hold times (dwell time)
 * Time between keydown and corresponding keyup
 */
function calculateKeyHoldTimes(keystrokesData) {
  const keyMap = new Map(); // Map of key to keydown timestamp

  const holdTimes = [];

  for (const event of keystrokesData) {
    if (event.type === 'keydown') {
      keyMap.set(event.key, event.timestamp);
    } else if (event.type === 'keyup' && keyMap.has(event.key)) {
      const downTime = keyMap.get(event.key);
      const holdTime = event.timestamp - downTime;
      if (holdTime >= 0) {
        holdTimes.push(holdTime);
      }
      keyMap.delete(event.key);
    }
  }

  return {
    values: holdTimes,
    mean: holdTimes.length > 0 ? holdTimes.reduce((a, b) => a + b, 0) / holdTimes.length : 0,
    median: calculateMedian(holdTimes),
    stdDev: calculateStdDev(holdTimes),
    min: holdTimes.length > 0 ? Math.min(...holdTimes) : 0,
    max: holdTimes.length > 0 ? Math.max(...holdTimes) : 0,
    count: holdTimes.length,
  };
}

/**
 * Calculate flight times (inter-keystroke intervals)
 * Time between keyup of one key and keydown of next
 */
function calculateFlightTimes(keystrokesData) {
  let lastUpTime = null;
  const flightTimes = [];

  for (const event of keystrokesData) {
    if (event.type === 'keydown' && lastUpTime !== null) {
      const flightTime = event.timestamp - lastUpTime;
      if (flightTime >= 0) {
        flightTimes.push(flightTime);
      }
    }
    if (event.type === 'keyup') {
      lastUpTime = event.timestamp;
    }
  }

  return {
    values: flightTimes,
    mean: flightTimes.length > 0 ? flightTimes.reduce((a, b) => a + b, 0) / flightTimes.length : 0,
    median: calculateMedian(flightTimes),
    stdDev: calculateStdDev(flightTimes),
    min: flightTimes.length > 0 ? Math.min(...flightTimes) : 0,
    max: flightTimes.length > 0 ? Math.max(...flightTimes) : 0,
    count: flightTimes.length,
  };
}

/**
 * Calculate typing speed metrics
 */
function calculateTypingSpeed(keystrokesData, typedText) {
  if (!keystrokesData || keystrokesData.length === 0) {
    return { wpm: 0, cpm: 0, duration: 0 };
  }

  const firstEvent = keystrokesData[0];
  const lastEvent = keystrokesData[keystrokesData.length - 1];
  const duration = (lastEvent.timestamp - firstEvent.timestamp) / 1000; // Convert to seconds

  const wordCount = typedText ? typedText.trim().split(/\s+/).length : 0;
  const charCount = typedText ? typedText.length : 0;

  return {
    wpm: duration > 0 ? ((wordCount / duration) * 60).toFixed(2) : 0,
    cpm: duration > 0 ? ((charCount / duration) * 60).toFixed(2) : 0,
    duration: duration.toFixed(2),
    keyCount: keystrokesData.length,
  };
}

/**
 * Calculate rhythm and regularity metrics
 */
function calculateRhythmMetrics(keystrokesData) {
  // Get all interval times (combined hold + flight times)
  const intervals = [];
  let lastTime = null;

  for (const event of keystrokesData) {
    if (lastTime !== null) {
      intervals.push(event.timestamp - lastTime);
    }
    lastTime = event.timestamp;
  }

  const stdDev = calculateStdDev(intervals);
  const mean = intervals.length > 0 ? intervals.reduce((a, b) => a + b, 0) / intervals.length : 0;

  // Rhythm regularity: inverse of coefficient of variation
  const coefficientOfVariation = mean > 0 ? (stdDev / mean) : 0;
  const rhythmRegularity = Math.max(0, (1 - coefficientOfVariation) * 100).toFixed(2); // 0-100%

  return {
    meanInterval: mean.toFixed(2),
    stdDevInterval: stdDev.toFixed(2),
    coefficientOfVariation: coefficientOfVariation.toFixed(3),
    rhythmRegularity: rhythmRegularity, // Higher = more regular
  };
}

/**
 * Analyze pause patterns
 * Pauses defined as intervals significantly longer than typical inter-keystroke time
 */
function analyzePauses(keystrokesData) {
  // Calculate inter-keystroke intervals
  const intervals = [];
  let lastTime = null;

  for (const event of keystrokesData) {
    if (lastTime !== null) {
      intervals.push(event.timestamp - lastTime);
    }
    lastTime = event.timestamp;
  }

  if (intervals.length === 0) {
    return {
      pauseCount: 0,
      pauseDurations: [],
      meanPauseDuration: 0,
      totalPauseTime: 0,
      pauseFrequency: 0,
    };
  }

  // Pause threshold: 1.5x the 75th percentile of intervals
  const sortedIntervals = [...intervals].sort((a, b) => a - b);
  const q75 = sortedIntervals[Math.floor(sortedIntervals.length * 0.75)];
  const pauseThreshold = q75 * 1.5;

  const pauseDurations = intervals.filter(i => i > pauseThreshold);
  const pauseCount = pauseDurations.length;
  const totalPauseTime = pauseDurations.reduce((a, b) => a + b, 0);
  const meanPauseDuration = pauseCount > 0 ? totalPauseTime / pauseCount : 0;

  return {
    pauseCount,
    pauseDurations,
    meanPauseDuration: meanPauseDuration.toFixed(2),
    totalPauseTime: totalPauseTime.toFixed(2),
    pauseFrequency: (pauseCount / intervals.length * 100).toFixed(2), // Percentage of intervals that are pauses
    pauseThreshold: pauseThreshold.toFixed(2),
  };
}

/**
 * Calculate error and correction features
 */
function calculateErrorMetrics(typedText, promptText, keystrokesData) {
  if (!typedText || !promptText) {
    return {
      totalErrors: 0,
      errorRate: 0,
      corrections: 0,
      errorPositions: [],
    };
  }

  // Character-level error detection
  const errors = [];
  const minLength = Math.min(typedText.length, promptText.length);

  for (let i = 0; i < minLength; i++) {
    if (typedText[i] !== promptText[i]) {
      errors.push(i);
    }
  }

  // Additional length-based errors
  const lengthDiff = Math.abs(typedText.length - promptText.length);
  const totalErrors = errors.length + lengthDiff;

  // Correction detection (simple heuristic: look for backspace events)
  const backspaceCount = keystrokesData.filter(k => k.key === 'Backspace').length;

  return {
    totalErrors,
    errorRate: ((totalErrors / promptText.length) * 100).toFixed(2),
    errorPositions: errors,
    corrections: backspaceCount,
    typingAccuracy: ((1 - totalErrors / promptText.length) * 100).toFixed(2),
  };
}

/**
 * Calculate keystroke statistics
 */
function calculateKeystrokeStats(keystrokesData, typedText) {
  const keydownCount = keystrokesData.filter(k => k.type === 'keydown').length;
  const keyupCount = keystrokesData.filter(k => k.type === 'keyup').length;

  // Special key analysis
  const shiftKeyCount = keystrokesData.filter(k => k.shiftKey).length;
  const ctrlKeyCount = keystrokesData.filter(k => k.ctrlKey).length;
  const backspaceCount = keystrokesData.filter(k => k.key === 'Backspace').length;
  const spaceCount = keystrokesData.filter(k => k.key === ' ').length;

  return {
    totalKeystrokes: keystrokesData.length,
    keydownCount,
    keyupCount,
    shiftKeyCount,
    ctrlKeyCount,
    backspaceCount,
    spaceCount,
    typedCharCount: typedText ? typedText.length : 0,
    uniqueKeys: new Set(keystrokesData.map(k => k.key)).size,
  };
}

/**
 * Normalize features relative to baseline
 * Baseline is typically the low cognitive load session or user's average
 */
export function normalizeFeatures(currentFeatures, baselineFeatures) {
  if (!baselineFeatures) {
    return currentFeatures; // Return unnormalized if no baseline
  }

  const normalized = {
    keyHoldTime: {
      meanNormalized: normalizeMetric(
        currentFeatures.keyHoldTimes.mean,
        baselineFeatures.keyHoldTimes.mean
      ),
      stdDevNormalized: normalizeMetric(
        currentFeatures.keyHoldTimes.stdDev,
        baselineFeatures.keyHoldTimes.stdDev
      ),
    },
    flightTime: {
      meanNormalized: normalizeMetric(
        currentFeatures.flightTimes.mean,
        baselineFeatures.flightTimes.mean
      ),
      stdDevNormalized: normalizeMetric(
        currentFeatures.flightTimes.stdDev,
        baselineFeatures.flightTimes.stdDev
      ),
    },
    typingSpeed: {
      wpmNormalized: normalizeMetric(
        parseFloat(currentFeatures.typingSpeed.wpm),
        parseFloat(baselineFeatures.typingSpeed.wpm)
      ),
      cpmNormalized: normalizeMetric(
        parseFloat(currentFeatures.typingSpeed.cpm),
        parseFloat(baselineFeatures.typingSpeed.cpm)
      ),
    },
    rhythm: {
      regularityNormalized: normalizeMetric(
        parseFloat(currentFeatures.rhythmMetrics.rhythmRegularity),
        parseFloat(baselineFeatures.rhythmMetrics.rhythmRegularity)
      ),
    },
    pauses: {
      pauseCountNormalized: normalizeMetric(
        currentFeatures.pauses.pauseCount,
        baselineFeatures.pauses.pauseCount
      ),
      pauseDurationNormalized: normalizeMetric(
        parseFloat(currentFeatures.pauses.meanPauseDuration),
        parseFloat(baselineFeatures.pauses.meanPauseDuration)
      ),
    },
    errors: {
      errorRateNormalized: normalizeMetric(
        parseFloat(currentFeatures.errorMetrics.errorRate),
        parseFloat(baselineFeatures.errorMetrics.errorRate)
      ),
      correctionNormalized: normalizeMetric(
        currentFeatures.errorMetrics.corrections,
        baselineFeatures.errorMetrics.corrections
      ),
    },
  };

  return normalized;
}

/**
 * Normalize a metric: (current - baseline) / baseline
 * Positive value = increase from baseline
 * Negative value = decrease from baseline
 * Returns percentage change
 */
function normalizeMetric(current, baseline) {
  if (baseline === 0 || baseline === null || baseline === undefined) {
    return 0;
  }
  const change = ((current - baseline) / baseline) * 100;
  return isFinite(change) ? parseFloat(change.toFixed(2)) : 0;
}

/**
 * Helper: Calculate median
 */
function calculateMedian(values) {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 !== 0
    ? sorted[mid]
    : (sorted[mid - 1] + sorted[mid]) / 2;
}

/**
 * Helper: Calculate standard deviation
 */
function calculateStdDev(values) {
  if (values.length === 0) return 0;
  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  const variance =
    values.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / values.length;
  return Math.sqrt(variance);
}

/**
 * Get empty features object (for initialization)
 */
function getEmptyFeatures() {
  return {
    keyHoldTimes: {
      values: [],
      mean: 0,
      median: 0,
      stdDev: 0,
      min: 0,
      max: 0,
      count: 0,
    },
    flightTimes: {
      values: [],
      mean: 0,
      median: 0,
      stdDev: 0,
      min: 0,
      max: 0,
      count: 0,
    },
    typingSpeed: { wpm: 0, cpm: 0, duration: 0, keyCount: 0 },
    rhythmMetrics: {
      meanInterval: 0,
      stdDevInterval: 0,
      coefficientOfVariation: 0,
      rhythmRegularity: 0,
    },
    pauses: {
      pauseCount: 0,
      pauseDurations: [],
      meanPauseDuration: 0,
      totalPauseTime: 0,
      pauseFrequency: 0,
    },
    errorMetrics: {
      totalErrors: 0,
      errorRate: 0,
      corrections: 0,
      errorPositions: [],
      typingAccuracy: 0,
    },
    keystrokeStats: {
      totalKeystrokes: 0,
      keydownCount: 0,
      keyupCount: 0,
      shiftKeyCount: 0,
      ctrlKeyCount: 0,
      backspaceCount: 0,
      spaceCount: 0,
      typedCharCount: 0,
      uniqueKeys: 0,
    },
  };
}

/**
 * Generate summary statistics from keystroke features
 */
export function generateFeatureSummary(features) {
  return {
    typingSpeed: {
      wpm: features.typingSpeed.wpm,
      cpm: features.typingSpeed.cpm,
    },
    keyTiming: {
      avgHoldTime: features.keyHoldTimes.mean.toFixed(0),
      avgFlightTime: features.flightTimes.mean.toFixed(0),
    },
    rhythm: {
      regularity: features.rhythmMetrics.rhythmRegularity,
      consistency: (100 - parseFloat(features.rhythmMetrics.coefficientOfVariation) * 100)
        .toFixed(2),
    },
    pauses: {
      count: features.pauses.pauseCount,
      avgDuration: features.pauses.meanPauseDuration,
      frequency: features.pauses.pauseFrequency,
    },
    accuracy: {
      typing: features.errorMetrics.typingAccuracy,
      errorRate: features.errorMetrics.errorRate,
      corrections: features.errorMetrics.corrections,
    },
  };
}

File: keystroke-data-collector/backend/keystroke_features.py
"""
Keystroke Dynamics Feature Extraction Module
Comprehensive keystroke feature analysis including:
- Key hold time and flight time metrics
- Typing speed and rhythm indicators
- Pause duration and frequency
- Error and correction-related features
"""

import statistics
from collections import defaultdict


def extract_keystroke_features(keystroke_data, typed_text, prompt_text):
    """
    Extract keystroke metrics from raw keystroke events
    
    Args:
        keystroke_data: List of keystroke events with timestamp
        typed_text: The actual text typed by user
        prompt_text: The target text to type
        
    Returns:
        Dictionary with comprehensive keystroke metrics
    """
    if not keystroke_data:
        return get_empty_features()
    
    metrics = {
        'keyHoldTimes': calculate_key_hold_times(keystroke_data),
        'flightTimes': calculate_flight_times(keystroke_data),
        'typingSpeed': calculate_typing_speed(keystroke_data, typed_text),
        'rhythmMetrics': calculate_rhythm_metrics(keystroke_data),
        'pauses': analyze_pauses(keystroke_data),
        'errorMetrics': calculate_error_metrics(typed_text, prompt_text, keystroke_data),
        'keystrokeStats': calculate_keystroke_stats(keystroke_data, typed_text),
    }
    
    return metrics


def calculate_key_hold_times(keystroke_data):
    """Calculate key hold times (dwell time) from keydown/keyup pairs"""
    key_map = {}  # Maps key to keydown timestamp
    hold_times = []
    
    for event in keystroke_data:
        if event.get('type') == 'keydown':
            key_map[event.get('key')] = event.get('timestamp', 0)
        elif event.get('type') == 'keyup' and event.get('key') in key_map:
            down_time = key_map[event.get('key')]
            hold_time = event.get('timestamp', 0) - down_time
            if hold_time >= 0:
                hold_times.append(hold_time)
            del key_map[event.get('key')]
    
    return {
        'values': hold_times,
        'mean': statistics.mean(hold_times) if hold_times else 0,
        'median': statistics.median(hold_times) if hold_times else 0,
        'stdDev': statistics.stdev(hold_times) if len(hold_times) > 1 else 0,
        'min': min(hold_times) if hold_times else 0,
        'max': max(hold_times) if hold_times else 0,
        'count': len(hold_times),
    }


def calculate_flight_times(keystroke_data):
    """Calculate flight times (inter-keystroke intervals) from keyup to keydown"""
    flight_times = []
    last_up_time = None
    
    for event in keystroke_data:
        if event.get('type') == 'keydown' and last_up_time is not None:
            flight_time = event.get('timestamp', 0) - last_up_time
            if flight_time >= 0:
                flight_times.append(flight_time)
        if event.get('type') == 'keyup':
            last_up_time = event.get('timestamp', 0)
    
    return {
        'values': flight_times,
        'mean': statistics.mean(flight_times) if flight_times else 0,
        'median': statistics.median(flight_times) if flight_times else 0,
        'stdDev': statistics.stdev(flight_times) if len(flight_times) > 1 else 0,
        'min': min(flight_times) if flight_times else 0,
        'max': max(flight_times) if flight_times else 0,
        'count': len(flight_times),
    }


def calculate_typing_speed(keystroke_data, typed_text):
    """Calculate typing speed metrics"""
    if not keystroke_data:
        return {'wpm': 0, 'cpm': 0, 'duration': 0}
    
    first_event = keystroke_data[0]
    last_event = keystroke_data[-1]
    duration = (last_event.get('timestamp', 0) - first_event.get('timestamp', 0)) / 1000  # Convert to seconds
    
    word_count = len(typed_text.split()) if typed_text else 0
    char_count = len(typed_text) if typed_text else 0
    
    return {
        'wpm': round((word_count / duration) * 60, 2) if duration > 0 else 0,
        'cpm': round((char_count / duration) * 60, 2) if duration > 0 else 0,
        'duration': round(duration, 2),
        'keyCount': len(keystroke_data),
    }


def calculate_rhythm_metrics(keystroke_data):
    """Calculate rhythm and regularity metrics"""
    intervals = []
    last_time = None
    
    for event in keystroke_data:
        if last_time is not None:
            intervals.append(event.get('timestamp', 0) - last_time)
        last_time = event.get('timestamp', 0)
    
    if not intervals:
        return {
            'meanInterval': 0,
            'stdDevInterval': 0,
            'coefficientOfVariation': 0,
            'rhythmRegularity': 0,
        }
    
    mean = statistics.mean(intervals)
    std_dev = statistics.stdev(intervals) if len(intervals) > 1 else 0
    
    # Rhythm regularity: inverse of coefficient of variation
    cv = (std_dev / mean) if mean > 0 else 0
    rhythm_regularity = max(0, (1 - cv) * 100)
    
    return {
        'meanInterval': round(mean, 2),
        'stdDevInterval': round(std_dev, 2),
        'coefficientOfVariation': round(cv, 3),
        'rhythmRegularity': round(rhythm_regularity, 2),
    }


def analyze_pauses(keystroke_data):
    """Analyze pause patterns in typing"""
    intervals = []
    last_time = None
    
    for event in keystroke_data:
        if last_time is not None:
            intervals.append(event.get('timestamp', 0) - last_time)
        last_time = event.get('timestamp', 0)
    
    if not intervals:
        return {
            'pauseCount': 0,
            'pauseDurations': [],
            'meanPauseDuration': 0,
            'totalPauseTime': 0,
            'pauseFrequency': 0,
        }
    
    # Pause threshold: 1.5x the 75th percentile
    sorted_intervals = sorted(intervals)
    q75_index = int(len(sorted_intervals) * 0.75)
    q75 = sorted_intervals[q75_index] if q75_index < len(sorted_intervals) else sorted_intervals[-1]
    pause_threshold = q75 * 1.5
    
    pause_durations = [i for i in intervals if i > pause_threshold]
    pause_count = len(pause_durations)
    total_pause_time = sum(pause_durations) if pause_durations else 0
    mean_pause = total_pause_time / pause_count if pause_count > 0 else 0
    pause_frequency = (pause_count / len(intervals) * 100) if intervals else 0
    
    return {
        'pauseCount': pause_count,
        'pauseDurations': pause_durations,
        'meanPauseDuration': round(mean_pause, 2),
        'totalPauseTime': round(total_pause_time, 2),
        'pauseFrequency': round(pause_frequency, 2),
        'pauseThreshold': round(pause_threshold, 2),
    }


def calculate_error_metrics(typed_text, prompt_text, keystroke_data):
    """Calculate error and correction features"""
    if not typed_text or not prompt_text:
        return {
            'totalErrors': 0,
            'errorRate': 0,
            'corrections': 0,
            'errorPositions': [],
            'typingAccuracy': 0,
        }
    
    # Character-level error detection
    errors = []
    min_length = min(len(typed_text), len(prompt_text))
    
    for i in range(min_length):
        if typed_text[i] != prompt_text[i]:
            errors.append(i)
    
    # Length-based errors
    length_diff = abs(len(typed_text) - len(prompt_text))
    total_errors = len(errors) + length_diff
    
    # Correction detection (backspace count)
    backspace_count = sum(1 for event in keystroke_data if event.get('key') == 'Backspace')
    
    error_rate = (total_errors / len(prompt_text) * 100) if prompt_text else 0
    typing_accuracy = (1 - total_errors / len(prompt_text) * 100) if prompt_text else 0
    
    return {
        'totalErrors': total_errors,
        'errorRate': round(error_rate, 2),
        'errorPositions': errors,
        'corrections': backspace_count,
        'typingAccuracy': round(max(0, typing_accuracy), 2),
    }


def calculate_keystroke_stats(keystroke_data, typed_text):
    """Calculate keystroke statistics"""
    keydown_count = sum(1 for event in keystroke_data if event.get('type') == 'keydown')
    keyup_count = sum(1 for event in keystroke_data if event.get('type') == 'keyup')
    
    shift_count = sum(1 for event in keystroke_data if event.get('shiftKey'))
    ctrl_count = sum(1 for event in keystroke_data if event.get('ctrlKey'))
    backspace_count = sum(1 for event in keystroke_data if event.get('key') == 'Backspace')
    space_count = sum(1 for event in keystroke_data if event.get('key') == ' ')
    
    unique_keys = len(set(event.get('key') for event in keystroke_data))
    
    return {
        'totalKeystrokes': len(keystroke_data),
        'keydownCount': keydown_count,
        'keyupCount': keyup_count,
        'shiftKeyCount': shift_count,
        'ctrlKeyCount': ctrl_count,
        'backspaceCount': backspace_count,
        'spaceCount': space_count,
        'typedCharCount': len(typed_text) if typed_text else 0,
        'uniqueKeys': unique_keys,
    }


def normalize_features(current_features, baseline_features):
    """
    Normalize features relative to baseline
    Returns percentage change from baseline
    """
    if not baseline_features:
        return current_features
    
    def normalize_metric(current, baseline):
        if baseline == 0 or baseline is None:
            return 0
        change = ((current - baseline) / baseline) * 100
        return round(change, 2) if isinstance(change, (int, float)) else 0
    
    normalized = {
        'keyHoldTime': {
            'meanNormalized': normalize_metric(
                current_features['keyHoldTimes']['mean'],
                baseline_features['keyHoldTimes']['mean']
            ),
            'stdDevNormalized': normalize_metric(
                current_features['keyHoldTimes']['stdDev'],
                baseline_features['keyHoldTimes']['stdDev']
            ),
        },
        'flightTime': {
            'meanNormalized': normalize_metric(
                current_features['flightTimes']['mean'],
                baseline_features['flightTimes']['mean']
            ),
            'stdDevNormalized': normalize_metric(
                current_features['flightTimes']['stdDev'],
                baseline_features['flightTimes']['stdDev']
            ),
        },
        'typingSpeed': {
            'wpmNormalized': normalize_metric(
                current_features['typingSpeed']['wpm'],
                baseline_features['typingSpeed']['wpm']
            ),
            'cpmNormalized': normalize_metric(
                current_features['typingSpeed']['cpm'],
                baseline_features['typingSpeed']['cpm']
            ),
        },
        'errors': {
            'errorRateNormalized': normalize_metric(
                current_features['errorMetrics']['errorRate'],
                baseline_features['errorMetrics']['errorRate']
            ),
            'correctionNormalized': normalize_metric(
                current_features['errorMetrics']['corrections'],
                baseline_features['errorMetrics']['corrections']
            ),
        },
    }
    
    return normalized


def get_empty_features():
    """Return empty features object for initialization"""
    return {
        'keyHoldTimes': {
            'values': [],
            'mean': 0,
            'median': 0,
            'stdDev': 0,
            'min': 0,
            'max': 0,
            'count': 0,
        },
        'flightTimes': {
            'values': [],
            'mean': 0,
            'median': 0,
            'stdDev': 0,
            'min': 0,
            'max': 0,
            'count': 0,
        },
        'typingSpeed': {
            'wpm': 0,
            'cpm': 0,
            'duration': 0,
            'keyCount': 0,
        },
        'rhythmMetrics': {
            'meanInterval': 0,
            'stdDevInterval': 0,
            'coefficientOfVariation': 0,
            'rhythmRegularity': 0,
        },
        'pauses': {
            'pauseCount': 0,
            'pauseDurations': [],
            'meanPauseDuration': 0,
            'totalPauseTime': 0,
            'pauseFrequency': 0,
        },
        'errorMetrics': {
            'totalErrors': 0,
            'errorRate': 0,
            'corrections': 0,
            'errorPositions': [],
            'typingAccuracy': 0,
        },
        'keystrokeStats': {
            'totalKeystrokes': 0,
            'keydownCount': 0,
            'keyupCount': 0,
            'shiftKeyCount': 0,
            'ctrlKeyCount': 0,
            'backspaceCount': 0,
            'spaceCount': 0,
            'typedCharCount': 0,
            'uniqueKeys': 0,
        },
    }
